src/Symfony/Component/Encryption/Ciphertext.php

```diff
@@ -0,0 +1,170 @@
+use Symfony\Component\Encryption\Exception\MalformedCipherException;
+
+/**
+ * This class is responsible over the payload API.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 5.3
+ *
+ * @internal
+ */
+class Ciphertext
+{
+    /**
+     * @var string algorithm for the encryption
+     */
+    private $algorithm;
+
+    /**
+     * @var string the encoded payload
+     */
+    private $payload;
+
+    /**
+     * @var string nonce for the algorithm
+     */
+    private $nonce;
+
+    /**
+     * @var array<string, string> additional headers
+     */
+    private $headers = [];
+
+    private function __construct()
+    {
+    }
+
+    public static function create(string $algorithm, string $ciphertext, string $nonce, array $headers = []): self
+    {
+        $model = new self();
+        $model->algorithm = $algorithm;
+        $model->payload = $ciphertext;
+        $model->nonce = $nonce;
+        $model->headers = $headers;
+
+        return $model;
+    }
+
+    /**
+     * Take a string representation of the chiphertext and parse it into an object.
+     *
+     * @throws MalformedCipherException
+     */
+    public static function parse(string $input): self
+    {
+        $parts = explode('.', $input);
+        if (!\is_array($parts) || 4 !== \count($parts)) {
+            throw new MalformedCipherException();
+        }
+
+        [$headersString, $payload, $nonce, $hashSignature] = $parts;
+
+        $headersString = self::base64UrlDecode($headersString);
+        $payload = self::base64UrlDecode($payload);
+        $nonce = self::base64UrlDecode($nonce);
+        $hashSignature = self::base64UrlDecode($hashSignature);
+
+        // Check if Authentication Tag is valid
+        $hash = hash('sha256', $headersString.$payload.$nonce);
+        if (!hash_equals($hash, $hashSignature)) {
+            throw new MalformedCipherException();
+        }
+
+        $headers = json_decode($headersString, true);
+        if (!\is_array($headers) || !\array_key_exists('alg', $headers)) {
+            throw new MalformedCipherException();
+        }
+
+        $model = new self();
+        $model->algorithm = $headers['alg'];
+        unset($headers['alg']);
+        $model->headers = $headers;
+        $model->nonce = $nonce;
+        $model->payload = $payload;
+
+        return $model;
+    }
+
+    /**
+     * @return string
+     */
+    public function __toString()
+    {
+        return $this->getString();
+    }
+
+    public function getString(): string
+    {
+        $headers = json_encode(array_merge($this->headers, [
+            'alg' => $this->algorithm,
+            'ver' => '1',
+        ]));
+
+        return sprintf('%s.%s.%s.%s',
+            self::base64UrlEncode($headers),
+            self::base64UrlEncode($this->payload),
+            self::base64UrlEncode($this->nonce),
+            self::base64UrlEncode(hash('sha256', $headers.$this->payload.$this->nonce))
```

Comment:
What is the purpose of this signature?

Reply:
It is an integrity check to make sure we have received the full string. 

This is less useful when you send a message over TCP but is more useful when you store the key on disk or in db. Basically a quick sanity check.

Reply:
I'm also wondering about this signature. Doesn't the encrypted payload already handle its own integrity?

Reply:
Yes, the payload does. But not the header and nonce.

Reply:
hmm. in that case, the same nonce is used twice. Isn't it a weakness?

do we really need this headers property? I feel the code and the api would be more simpler without it. 
people can json_encode their own array if they want providing extra data